### PR TITLE
New fileset for googlecloud firewall logs

### PR DESCRIPTION
**Pull request by adriansr**
This PR adds a new fileset, `firewall`, to the `googlecloud` module in Filebeat. It helps parsing firewall logs generated by rules under _VPC Network -> Firewall Rules._ 

Note that GCP only logs firewall events under the following conditions:
- Logging needs to be enabled for each individual rule in order to log.
- Only TCP and UDP rules can be logged.

See https://cloud.google.com/vpc/docs/using-firewall-rules-logging.

### CODE REVIEW

File: x-pack/filebeat/module/googlecloud/firewall/_meta/fields.yml [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+      - name: source_service_account
+        type: keyword
+        description: >
+          List of all the source service accounts that the firewall rule applies to.
+      - name: target_service_account
+        type: keyword
+        description: >
+          List of all the target service accounts that the firewall rule applies to.
+
+  - name: destination.instance
```
Comment (adriansr) [inline x-pack/filebeat/module/googlecloud/firewall/_meta/fields.yml]: This group and some other below are duplicated from vpcflows. I think it's worth to move them to `googlecloud.destination.instance` etc. but I'm unsure about backwards-compatibility. Is setting the previous `googlecloud.vpcflow.xxx` field as an alias enough and safe?

Comment (andrewkroh) [reply x-pack/filebeat/module/googlecloud/firewall/_meta/fields.yml]: I think moving them up to the module level would be good. The vpcflow fileset is still in beta so we should be able to move those fields without adding the alias.

File: x-pack/filebeat/module/googlecloud/firewall/config/input.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+{{ end }}
+
+processors:
+  - script:
+      lang: javascript
+      id: googlecloud_firewall_script
+      params:
+        debug: {{ .debug }}
+        keep_original: {{ .keep_original_message }}
```
Comment (andrewkroh) [inline x-pack/filebeat/module/googlecloud/firewall/config/input.yml]: Consider using `keep_original_message` as the parameter name to the module.

File: x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+// makeConditional({condition:expr, result1:processor|expr, [...]})
+//
+// Processor that selects which processor to run depending on the result of
+// evaluating a _condition_. Result can be boolean (if-else equivalent) or any
+// other value (switch equivalent). Unspecified values are a no-op.
+function makeConditional(options) {
+    return function (evt) {
+        var branch = options[options.condition(evt)] || function(evt){};
+        return (typeof branch === "function"? branch : branch.Run)(evt);
```
Comment (andrewkroh) [inline x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js]:
```suggestion
        return (typeof branch === "function" ? branch : branch.Run)(evt);
```

File: x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+//
+// Processor that logs the current value of evt to console.debug.
+function makeLogEvent(msg) {
+    return function (evt) {
+        console.debug(msg + " :" +  JSON.stringify(evt, null, 4));
+    };
+}
+
+// PipelineBuilder to aid debugging of pipelines during development.
+function PipelineBuilder(pipelineName, debug) {
```
Comment (andrewkroh) [inline x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js]: Nice idea.

File: x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+        target: "event",
+        fields: {
+            category: "firewall-rule",
+            type: "firewall"
+        },
+    }));
+
+    builder.Add("saveMetadata", new processor.Convert({
+        fields: [
+            {from: "json.logName", to:"log.logger"},
```
Comment (andrewkroh) [inline x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js]:
```suggestion
            {from: "json.logName", to: "log.logger"},
```

File: x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+        fields: [
+            {from: "json.jsonPayload", to: "json"},
+        ],
+        mode: "rename"
+    }));
+
+    builder.Add("addOutcome", makeMapper({
+        from: "json.disposition",
+        to: "event.outcome",
+        mappings: {
```
Comment (andrewkroh) [inline x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js]: Another nice pattern. 😄

File: x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+    var chain = builder.Build();
+    return {
+        process: chain.Run
+    };
+}
+
+var firewall;
+
+function process(evt) {
+    if (firewall == null) {
```
Comment (andrewkroh) [inline x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js]: Could this be done in the `register()` call?

File: x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js [hunk 8]
```diff
@@ -0,0 +1,331 @@
...
+    }));
+
+    builder.Add("addOutcome", makeMapper({
+        from: "json.disposition",
+        to: "event.outcome",
+        mappings: {
+            ALLOWED: "allow",
+            DENIED: "deny"
+        },
+        default: "unknown"
```
Comment (webmat) [inline x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js]: Please note that `event.outcome` is still reserved.

The current thinking is that expected values in this field (if present) should be: "success" or "failure". Finer grained information such as "allow" or "deny" should be in another place, perhaps in `event.action` or a custom field.

As usual, populating a reserved field signs you up to having to do a breaking change later.

Comment (webmat) [reply x-pack/filebeat/module/googlecloud/firewall/config/pipeline.js]: cc @MikePaquette

File: x-pack/filebeat/module/googlecloud/firewall/test/rare.log-expected.json [hunk 9]
```diff
@@ -0,0 +1,120 @@
+[
+    {
+        "@timestamp": "2019-11-06T16:41:38.394Z",
+        "destination.address": "10.128.0.16",
+        "destination.domain": "local-adrian-test",
+        "destination.ip": "10.128.0.16",
+        "destination.port": 80,
+        "event.category": "firewall-rule",
```
Comment (webmat) [inline x-pack/filebeat/module/googlecloud/firewall/test/rare.log-expected.json]: `event.category` is still reserved, and at this time, there is no plan to have a category named "firewall-rule".

Leaving this here signs you up to have to do a breaking change, once the expected values are published for `event.category`. The name of the category that will encapsulate firewall rule events is still very much in flux.

cc @MikePaquette

File: x-pack/filebeat/module/googlecloud/firewall/test/test.log-expected.json [hunk 10]
```diff
@@ -0,0 +1,1187 @@
...
+        "network.community_id": "1:iiDdIEXnxwSiz/hJbVnseQ4SZVE=",
+        "network.direction": "outbound",
+        "network.iana_number": 17,
+        "network.name": "default",
+        "network.transport": "udp",
+        "network.type": "ipv4",
+        "related.ip": [
+            "10.128.0.16",
+            "8.8.8.8"
+        ],
```
Comment (webmat) [inline x-pack/filebeat/module/googlecloud/firewall/test/test.log-expected.json]: OMG thanks for filling that ❤️ 🙂

**Comment by webmat** [comment]
Other than me harping on the reserved fields, I'm really excited about this! Good work, as usual :-)

**Comment by adriansr** [comment]
Thanks @webmat

Regarding `event.outcome` / `event.category`, those are the fields and values already in use by other firewall data sources (cisco, iptables, panos), so I think its better to keep compatibility now and when those fields are standarized we can rename them all.

**Pull request merged**